### PR TITLE
Make loss calculation deterministic and sync-free

### PR DESCRIPTION
**Pull request by ndryden**
* It turns out `reduction=sum` is nondeterministic. Switching to no reduction and following it with a `.sum()` is deterministic and costs little.
* Our use of `bincount` required a host/device synchronization. This has been replaced.
* Small unrelated comment updates.

Code by Claude.

### CODE REVIEW

File: ScaFFold/utils/losses.py [hunk 1]
```diff
@@ -146,25 +146,49 @@ def compute_sharded_cross_entropy_loss(
...
+        # enough to perturb train_stats.csv and to flip a best-checkpoint tie).
+        # It has no deterministic implementation at all: under
+        # torch.use_deterministic_algorithms(True) the fused form raises, and
+        # `more_determinism` passes warn_only=True, so it warns and stays
+        # nondeterministic. The separate .sum() is an ordinary tree reduction
+        # over a fixed shape and is bitwise reproducible.
+        #
+        # It is also not a cost at the scale that matters. Measured here,
+        # fwd+bwd, 7 classes, paired alternating arms:
+        #     128**3 (scale 7)  241 -> 147 us   0.61x -- the split form wins,
```
Comment (michaelmckinsey1) [inline ScaFFold/utils/losses.py]: same issue with too much information as #98

**Review by michaelmckinsey1 — APPROVED** [review]
I think we should make the same edits that we discussed to docstrings/comments in #98 uniformly on all of these PRs

**Review by PatrickRMiles — APPROVED** [review]
I agree with Michael that we should pare down the verbose comments. Some of these comments are helpful for understanding the proposed changes, like the comment about doing reduction="none" and summing after the F.nll_loss()/F.cross_entropy call to improve determinism, but I think this information should go into the commit message for this change instead.
